overlord/registrystate/registrymgr.go

```diff
@@ -37,3 +63,119 @@ func setupRegistryHook(st *state.State, snapName, hookName string, ignoreError b
 	task := hookstate.HookTask(st, summary, hookSup, nil)
 	return task
 }
+
+type changeViewHandler struct {
+	ctx *hookstate.Context
+}
+
+func (h *changeViewHandler) Before() error                 { return nil }
+func (h *changeViewHandler) Error(err error) (bool, error) { return false, nil }
+
+func (h *changeViewHandler) Done() error {
+	h.ctx.Lock()
+	defer h.ctx.Unlock()
+
+	t, _ := h.ctx.Task()
+	tx, _, err := GetStoredTransaction(t)
+	if err != nil {
+		return fmt.Errorf("cannot get transaction in change-registry handler: %v", err)
+	}
+
+	if tx.aborted() {
+		return fmt.Errorf("cannot change registry %s/%s: %s rejected changes: %s", tx.RegistryAccount, tx.RegistryName, tx.abortingSnap, tx.abortReason)
+	}
+
+	return nil
+}
+
+type saveViewHandler struct {
+	ctx *hookstate.Context
+}
+
+func (h *saveViewHandler) Before() error { return nil }
+
+func (h *saveViewHandler) Error(origErr error) (ignoreErr bool, err error) {
+	h.ctx.Lock()
+	defer h.ctx.Unlock()
+
+	t, _ := h.ctx.Task()
+	st := h.ctx.State()
+
+	var commitTaskID string
+	if err := t.Get("commit-task", &commitTaskID); err != nil {
+		return false, err
+	}
+
+	// stop all tasks (we need to do this manually because we're not erroring yet)
+	for _, t := range t.Change().Tasks() {
```

Comment:
this doesn't seem correct if the Change is doing more than just dealing with the transaction

Reply:
One alternative I considered was just creating the rollbacks tasks and putting the HaltTasks after those (which would mean they never run since the last rollback task surfaces the error). I'll move to that solution